### PR TITLE
network,passt: consider overhead when binding all ports

### PR DESCRIPTION
**Pull request by rhrazdil**
When VMI uses Passt interface and doesn't specify ports, all the ports are forwarded by Passt process.
Opening all the sockets for forwarding causes kernel to consume considerable amount of memory.

At the moment, Passt binding can only be used on primary interface, but since we need to loop over the VMI interfaces anyway, there's no reason not to be generic and account for each Passt binding with no specified Ports.

Signed-off-by: Radim Hrazdil <rhrazdil@redhat.com>

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
Add overhead of interface with Passt binding when no ports are specified
```

### CODE REVIEW

**Review by xpivarc — COMMENTED** [review]
/approve
/hold
Can we also re-evaluate passt tests that now add too much memory?

**Comment by rhrazdil** [comment]
> Can we also re-evaluate passt tests that now add too much memory?

Yep, that's a good point. Users shouldn't be required to do that with this change.

**Comment by AlonaKaplan** [comment]
Thanks!
/lgtm
/hold cancel

Please send a PR updating the [user guide](https://github.com/kubevirt/user-guide/blame/main/docs/virtual_machines/interfaces_and_networks.md#L646) as well.

**Comment by rhrazdil** [comment]
> Please send a PR updating the [user guide](https://github.com/kubevirt/user-guide/blame/main/docs/virtual_machines/interfaces_and_networks.md#L646) as well.

Will do!

**Comment by sbrivio-rh** [comment]
By the way, we consider that kind of memory usage an issue (of course), tracked at: https://bugs.passt.top/show_bug.cgi?id=20. I'll update that soon. Sneak preview: we're working on single per-port sockets for both IPv4 and IPv6 (Linux is kind of unique in supporting that), cutting down to ~ 350 MiB in worst case -- that's pretty much what we can do in userspace alone, and in parallel we're considering possible kernel improvements/extensions (where much larger gains are possible/expected).

File: pkg/virt-controller/services/renderresources.go [hunk 1]
```diff
@@ -357,6 +357,14 @@ func GetMemoryOverhead(vmi *v1.VirtualMachineInstance, cpuArch string) *resource
 		overhead.Add(resource.MustParse("53Mi"))
 	}
 
+	// Additional overhead for each interface with Passt binding, that forwards all ports.
+	// More information can be found here: https://bugs.passt.top/show_bug.cgi?id=20
+	for _, net := range vmi.Spec.Domain.Devices.Interfaces {
+		if net.Passt != nil && len(net.Ports) == 0 {
```
Comment (sbrivio-rh) [inline pkg/virt-controller/services/renderresources.go:363]: I'm not sure if you have a way to say that here (I wouldn't spend too long on it anyway as this should be temporary), but if it's IPv4-only or IPv6-only you can halve this.

Comment (rhrazdil) [reply pkg/virt-controller/services/renderresources.go:363]: There's no way to limit the binding to only IPv4 or IPv6 in the kubevirt Interface API.

Comment (sbrivio-rh) [reply pkg/virt-controller/services/renderresources.go:363]: And I missed that you already added the link to the ticket there. I could actually read commits before commenting :)

Comment (rhrazdil) [reply pkg/virt-controller/services/renderresources.go:363]: No problem. I'll open an issue for us to make sure we revise the memory overhead once the bug is closed.
It's a substantial memory reduction that you're working on, so it's important for us to no forget to take advantage of it.
Thanks for pointing that out!

**Pull request merged**